feat (Nodes): type `tslFn` and several functions that use tslFn

0b5vr opened the pull request:
### What

Type `tslFn` and several functions that use tslFn.

Not exhaustive.
Have very low confidence whether the typing direction is correct.

- `tslFn`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/shadernode/ShaderNode.js#L485-L520
- `BlendModeNode.js`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/display/BlendModeNode.js
- `BRDF_GGX`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_GGX.js
- `BRDF_Lambert`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_Lambert.js
- `BRDF_Sheen`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_Sheen.js
- `D_GGX`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/D_GGX.js
- `DFGApprox`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/DFGApprox.js
- `F_Schlick`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/F_Schlick.js

0b5vr committed:
feat (Nodes): type `tslFn` and several functions that uses tslFn

Not exhaustive. I rather want to check whether the direction is correct

- `tslFn`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/shadernode/ShaderNode.js#L485-L520
- `BlendModeNode.js`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/display/BlendModeNode.js
- `BRDF_GGX`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_GGX.js
- `BRDF_Lambert': https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_Lambert.js
- `BRDF_Sheen`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/BRDF_Sheen.js
- `D_GGX`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/D_GGX.js
- `DFGApprox`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/DFGApprox.js
- `F_Schlick`: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/functions/BSDF/F_Schlick.js

0b5vr committed:
refactor: `pnpm format`

Methuselah96 commented on types/three/examples/jsm/nodes/shadernode/ShaderNode.d.ts:
```diff
@@ -163,6 +163,10 @@ export function nodeImmutable<T>(
 
+export function tslFn<T, R extends Node = ShaderNodeObject<Node>>(
+    jsFunc: (args: T) => R,
+): T extends AnyObject ? (args: T) => R : () => R;
```
Can you explain the return type? I see that `tslFn` does fancy things with the first argument, but not seeing a clear explanation for why we would just remove the first argument if it's not an object. I would be inclined to just return `(args: T) => R` unless I'm missing something.

0b5vr replied:
`tslFn` can define functions without arguments, and it's used in the way in several nodes. `GaussianBlurNode` for example.

Definition: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/display/GaussianBlurNode.js#L103
Use: https://github.com/mrdoob/three.js/blob/r160/examples/jsm/nodes/display/GaussianBlurNode.js#L136

In this case, the user wants to call the function without argument, and the ternary return type achieves such cases.

To clarify the use, changing `T` to `T extends AnyObject | unknown` might help.

Methuselah96 replied:
Thanks, that's helpful. Can you add simple tests to `test/unit/examples/jsm/nodes/shaderNode/ShaderNode.ts` for both the single argument case and the no argument case? I feel like there might be a better way to type this, but I want to see this way functioning first before suggesting alternatives.

Methuselah96 replied:
One idea that crossed my mind is using function overloads instead. Something like:

```ts
export function tslFn<T, R extends Node = ShaderNodeObject<Node>>(
    jsFunc: (args: T) => R,
): (args: T) => R;
export function tslFn<R extends Node = ShaderNodeObject<Node>>(
    jsFunc: () => R,
): () => R;
```

Another idea might be to make the function itself the generic, but that might make it harder to manipulate the argument if needed.

0b5vr replied:
overloading looks great. I'll go with it. Thank you for the suggestion!

0b5vr replied:
added a test as well

0b5vr committed:
refactor: Rewrite tslFn using overloading

Based on a comment by @Methuselah96
https://github.com/three-types/three-ts-types/pull/735#discussion_r1452407532

0b5vr committed:
test: add test cases for tslFn

See: https://github.com/three-types/three-ts-types/pull/735#discussion_r1452404131

0b5vr committed:
refactor: pnpm format

Methuselah96 reviewed (approved):
Thanks!
